1. Apply various Machine Learning Models on Heart disease dataset and select the best suitable Model for given dataset

In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

In [4]:
colnames = ['age','sex','cp','trestbps','chol','fbs','restecg','thalach','exang','oldpeak','slope','ca','thal','num']
df = pd.read_csv("./Data/processed.cleveland.data",sep=',',names=colnames)

In [5]:
df

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0    63.0  1.0  1.0     145.0  233.0  1.0      2.0    150.0    0.0      2.3   
1    67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
2    67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   
3    37.0  1.0  3.0     130.0  250.0  0.0      0.0    187.0    0.0      3.5   
4    41.0  0.0  2.0     130.0  204.0  0.0      2.0    172.0    0.0      1.4   
..    ...  ...  ...       ...    ...  ...      ...      ...    ...      ...   
292  57.0  0.0  4.0     140.0  241.0  0.0      0.0    123.0    1.0      0.2   
293  45.0  1.0  1.0     110.0  264.0  0.0      0.0    132.0    0.0      1.2   
294  68.0  1.0  4.0     144.0  193.0  1.0      0.0    141.0    0.0      3.4   
295  57.0  1.0  4.0     130.0  131.0  0.0      0.0    115.0    1.0      1.2   
296  57.0  0.0  2.0     130.0  236.0  0.0      2.0    174.0    0.0      0.0   

     slope   ca  thal  num  
0      3.0  0.0   6.0    0  
1      2.0  3.0   3.0    2  
2      2.0  2.0   7.0    1  
3      3.0  0.0   3.0    0  
4      1.0  0.0   3.0    0  
..     ...  ...   ...  ...  
292    2.0  0.0   7.0    1  
293    2.0  0.0   7.0    1  
294    2.0  2.0   7.0    2  
295    2.0  1.0   7.0    3  
296    2.0  1.0   3.0    1  

[297 rows x 14 columns]

In [6]:
df['num'].value_counts()

num
0    160
1     54
2     35
3     35
4     13
Name: count, dtype: int64

In [7]:
df['num'].unique()

array([0, 2, 1, 3, 4])

In [8]:
# target column - num and values - 0,1,2,3,4 ,but I want to take 2 prosibilities - 0 and 1 
# 0 = no disease and 1,2,3,4 - convert 1 and 1 is disease 
# 1,2,3,4 convert 1 ,use of map function

df['num'] = df['num'].map({0:0,1:1,2:1,3:1,4:1})

In [9]:
x = df.iloc[:,:-1].values
y = df.iloc[:,-1].values

In [10]:
# train test split with 70% train and test 30%

from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=0)

In [11]:
# scale the data using Standard Scaler

from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train_1 = sc.fit_transform(x_train)
x_test_1 = sc.transform(x_test)

In [26]:
# Create Support Vector Classifier with train and test score

from sklearn.svm import SVC
svc = SVC(kernel='rbf')
svc.fit(x_train_1,y_train)
test_score_svc = svc.score(x_test_1,y_test)
train_score_svc = svc.score(x_train_1,y_train)
print("Train Score : %.2f" % train_score_svc)
print("Test Score: %.2f" % test_score_svc)



Train Score : 0.91
Test Score: 0.84


In [27]:
# Create Naive Bayes with train and test score

from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()
nb.fit(x_train,y_train)  # original data 
test_score_nb = nb.score(x_test,y_test)
train_score_nb = nb.score(x_train,y_train)
print("Train Score : %.2f" % train_score_nb)
print("Test Score: %.2f" % test_score_nb)


Train Score : 0.86
Test Score: 0.84


In [28]:
# Create Logistic Regression Model with train and test score

from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(x_train,y_train)
test_score_lr = lr.score(x_test,y_test)
train_score_lr = lr.score(x_train,y_train)
print("Train Score : %.2f" % train_score_lr)
print("Test Score: %.2f" % test_score_lr)


Train Score : 0.87
Test Score: 0.84


In [29]:
# Create Decision Tree Classifier Model with train and test score

from sklearn.tree import DecisionTreeClassifier
dtc = DecisionTreeClassifier()
dtc.fit(x_train,y_train)
test_score_dt = dtc.score(x_test,y_test)
train_score_dt = dtc.score(x_train,y_train)
print("Train Score : %.2f" % train_score_dt)
print("Test Score: %.2f" % test_score_dt)


Train Score : 1.00
Test Score: 0.74


In [41]:
# Create Random Forest  with train and test score

from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(x_train,y_train)
test_score_rf = rfc.score(x_test,y_test)
train_score_rf= rfc.score(x_train,y_train)
print("Train Score : %.2f" % train_score_rf)
print("Test Score: %.2f" % test_score_rf)

Train Score : 1.00
Test Score: 0.79


In [56]:
# create data frame with devaition and round all score

df_score = pd.DataFrame({'Model':['SVC','Naive Bayes','Logistic Regression','Decision Tree','Random Forest'],
                         'Train Score':[train_score_svc,train_score_nb,train_score_lr,train_score_dt,train_score_rf],
                         'Test Score':[test_score_svc,test_score_nb,test_score_lr,test_score_dt,test_score_rf]})
# round all score and  all score multiply by 100
df_score = df_score.round(2)
df_score['Train Score'] = df_score['Train Score']*100
df_score['Test Score'] = df_score['Test Score']*100
df_score

# difference between train score and test score add dataframe with name of deviation

df_score['Deviation'] =  df_score['Train Score'] - df_score['Test Score']
df_score

Model  Train Score  Test Score  Deviation
0                  SVC         91.0        84.0        7.0
1          Naive Bayes         86.0        84.0        2.0
2  Logistic Regression         87.0        84.0        3.0
3        Decision Tree        100.0        74.0       26.0
4        Random Forest        100.0        79.0       21.0

- Highest Training score is obtained in case of DT and RF but the deviation between train and test score is high .This indicates high variance or Overfitting the Model.
- The next best score is obtained with SVC where the highest score is 91% is reasonably good. In any ML Model,more than 90% is acceptable.There is not much deviation between train and test score.(In Deep Learning Models constructed with ANN,the acceptable score ranges from 70% onwards)
- SVC Model is the final model in this Dataset.